crates/sui-framework/sources/display.move

```diff
@@ -0,0 +1,186 @@
+    ///
+    /// Also, the only place it can be used is the function where the Display
+    /// object was created; hence values and names are likely to be hardcoded and
+    /// vector<u8> is the best type for that purpose.
+    public fun set_owned<T: key>(d: Display<T>, name: vector<u8>, value: vector<u8>): Display<T> {
+        set_internal(&mut d, utf8(name), utf8(value));
+        d
+    }
+
+    /// Create an empty Display object. It can either be shared empty of filled
+    /// with data right away via cheaper `set_owned` method.
+    public fun empty<T: key>(pub: &Publisher, ctx: &mut TxContext): Display<T> {
+        assert!(is_package<T>(pub), ENotOwner);
+
+        let uid = object::new(ctx);
+
+        event::emit(DisplayCreated<T> {
+            id: object::uid_to_inner(&uid)
+        });
+
+        Display {
+            id: uid,
+            fields: vec_map::empty()
+        }
+    }
+
+    /// Share an object after the initialization is complete.
+    public fun share<T: key>(d: Display<T>) {
+        transfer::share_object(d);
+    }
+
+    // ========== Entry functions ==========
+
+    /// Create a new empty Display<T> object and share it.
+    entry public fun create_and_share<T: key>(pub: &Publisher, ctx: &mut TxContext) {
+        share(empty<T>(pub, ctx))
+    }
+
+    /// Create a new Display<T> object with a set of fields.
+    entry public fun create_with_fields<T: key>(
+        pub: &Publisher, fields: vector<String>, values: vector<String>, ctx: &mut TxContext
+    ) {
+        let len = vector::length(&fields);
+        assert!(len == vector::length(&values), EVecLengthMismatch);
+
+        let i = 0;
+        let display = empty<T>(pub, ctx);
+        while (i < len) {
+            set_internal(&mut display, *vector::borrow(&fields, i), *vector::borrow(&values, i));
+            i = i + 1;
+        };
+
+        share(display)
+    }
+
+    /// Sets a custom `name` field with the `value`.
+    entry public fun set<T: key>(pub: &Publisher, d: &mut Display<T>, name: String, value: String) {
+        assert!(is_package<T>(pub), ENotOwner);
+        set_internal(d, name, value)
+    }
+
+    /// Sets multiple `fields` with `values`.
+    entry public fun set_multiple<T: key>(
+        pub: &Publisher, d: &mut Display<T>, fields: vector<String>, values: vector<String>
+    ) {
+        let len = vector::length(&fields);
+        assert!(is_package<T>(pub), ENotOwner);
+        assert!(len == vector::length(&values), EVecLengthMismatch);
+
+        let i = 0;
+        while (i < 0) {
+            set_internal(d, *vector::borrow(&fields, i), *vector::borrow(&values, i));
+            i = i + 1;
+        };
+    }
+
+    /// Private method for inserting fields without security checks.
+    fun set_internal<T: key>(d: &mut Display<T>, name: String, value: String) {
+        vec_map::insert(&mut d.fields, name, value)
+    }
+}
+
+#[test_only]
+module sui::display_capy {
```

Comment:
nit, aren't we doing separate files for tests?

Reply:
And I think `display_tests` would make this a bit easier to find

Reply:
Hahahaha